lib/impure/db_sqlite.nim

```diff
@@ -716,8 +719,64 @@ proc setEncoding*(connection: DbConn, encoding: string): bool {.
 proc finalize*(sqlPrepared:SqlPrepared){.discardable.} = 
   discard finalize(sqlPrepared.PStmt)
 
+template dbBindParamError*(paramIdx: int,val:untyped) =
+  ## Raises a `DbError` exception.
+  var e: ref DbError
+  new(e)
+  e.msg = "error binding param in position " & $paramIdx 
+  raise e
+
+proc bindParam*(ps: SqlPrepared, paramIdx: int, val: int32) =
+  ## Binds a int32  to the specified paramIndex.
+  if SQLITE_OK != bind_int(ps.PStmt, paramIdx.int32, val):
+    dbBindParamError(paramIdx,val)
+
+proc bindParam*(ps: SqlPrepared, paramIdx: int, val: int) =
+  ## Binds a int32  to the specified paramIndex.
+  if SQLITE_OK != bind_int(ps.PStmt, paramIdx.int32, val.int32):
+    dbBindParamError(paramIdx,val)
+
+proc bindParam*(ps: SqlPrepared, paramIdx: int,val: int64) =
+  ## Binds a int64  to the specified paramIndex.
+  if SQLITE_OK != bind_int64(ps.PStmt, paramIdx.int32, val):
+    dbBindParamError(paramIdx,val)
+
+proc bindParam*(ps: SqlPrepared, paramIdx: int, val: float64) =
+  ## Binds a 64bit float to the specified paramIndex.
+  if SQLITE_OK != bind_double(ps.PStmt, paramIdx.int32, val):
+    dbBindParamError(paramIdx,val)
+
+proc bindNull*(ps: SqlPrepared, paramIdx: int) =
+  ## Sets the bindparam at the specified paramIndex to null 
+  ## (default behaviour by sqlite).
+  if SQLITE_OK != bind_null(ps.PStmt, paramIdx.int32):
+    dbBindParamError(paramIdx,val)
+
+proc bindParam*(ps: SqlPrepared, paramIdx: int,val: string) =
+  ## Binds a string to the specified paramIndex.
+  if SQLITE_OK != bind_text(ps.PStmt, paramIdx.int32,val.cstring,-1.int32 , SQLITE_STATIC):
+    dbBindParamError(paramIdx,val)
+
+proc bindParam*(ps: SqlPrepared, paramIdx: int,val: cstring) =
+  ## binds a blob to the specified paramIndex.
+  let len = val.len
+  if SQLITE_OK != bind_blob(ps.PStmt, paramIdx.int32, val[0].unSafeAddr, len.int32 , SQLITE_STATIC):
+    dbBindParamError(paramIdx,val)
+
+template bindParams*(ps: SqlPrepared, params:untyped) = 
+  var idx:int = 1
+  when params is tuple or params is object:
+    for v in params.fields:
+      ps.bindParam(idx,v)
+      inc idx
+  else:
+    for v in params:
+      ps.bindParam(idx,v)
+      inc idx
```

Comment:
This template limits the current use case, since it forces you to use a tuple and/or objects when you want a set of differently typed arguments.

```nim
ps.bindParams((0, 42.0, "string"))
ps.bindParams((value: 0, str: "foo"))
```

This can also cause confusion if users just unknowingly bind an object thinking that it will bind the object as one parameter.

Instead if we use a macro like this:

```nim
macro bindParams(ps: SqlPrepared, params: varargs[untyped]): untyped =
  let bindParam = bindSym("bindParam", brOpen)
  let preparedStatement = genSym()
  result = newStmtList()
  # Store `ps` in a temporary variable. This prevents `ps` from being evaluated every call.
  result.add newNimNode(nnkLetSection).add(newIdentDefs(preparedStatement, newEmptyNode(), ps))
  for idx, param in params:
    result.add newCall(bindParam, ps, newIntLit idx, param)
```

Then binding will simply use the typically expected API of:

```nim
ps.bindParams(1, 10, 10000, "string", 42.0)
```

Not to mention that it will transparently work with the case where only one argument is used:

```nim
ps.bindParams("string")
```

Reply:
hmm, I thought db c api support bind param by index or name, so tuple and object ,array, seq are good choice. when `params` becoming  `varargs[untyped]` I dont know how to handle unexpected objects. passing user defined type object just ignored.

Reply:
> I dont know how to handle unexpected objects.

If no `bindParam` proc was defined for a particular type, this would simply error at compile time, same as with how your code works atm.

> db c api support bind param by index or name.

Then we should expose single-param versions of those, then we can think of using something like `openArray[(string, untyped)]` to enable the `{"param1": 10, "foo": "bar"}` syntax. The JSON module already did this, so we can learn from the logic there.